File: Corte1/C++/Evaluacion/parcial.cpp
#include <iostream>
#include <string>
using namespace std;

const int MAX_ESTUDIANTES = 30;
const int NUM_NOTAS = 3;


// VALIDACION DATOS

int leerEntero() {
    int valor;
    while (true) {
        cin >> valor;

        if (cin.fail()) {
            cin.clear();
            cin.ignore(1000, '\n');
            cout << "Entrada invalida. Ingrese un numero: ";
        } else {
            cin.ignore(1000, '\n');
            return valor;
        }
    }
}

float leerFloat() {
    float valor;
    while (true) {
        cin >> valor;

        if (cin.fail()) {
            cin.clear();
            cin.ignore(1000, '\n');
            cout << "Entrada invalida. Ingrese un numero: ";
        } else {
            cin.ignore(1000, '\n');
            return valor;
        }
    }
}

string leerTexto() {
    string texto;
    getline(cin, texto);

    while (texto.empty()) {
        cout << "No puede estar vacio. Intente de nuevo: ";
        getline(cin, texto);
    }

    return texto;
}


// ESTUDIANTE

class Estudiante {
private:
    int codigo;
    string nombre;
    int edad;

public:
    Estudiante() {
        codigo = 0;
        nombre = "";
        edad = 0;
    }

    Estudiante(int c, string n, int e) {
        codigo = c;
        nombre = n;
        edad = e;
    }

    int getCodigo() { 
	return codigo; }
    string getNombre() { 
	return nombre; }
    int getEdad() { 
	return edad; }

    void mostrar() {
        cout << "Codigo: " << codigo << endl;
        cout << "Nombre: " << nombre << endl;
        cout << "Edad: " << edad << endl;
    }
};


// SISTEMA

class Sistema {
private:
    Estudiante estudiantes[MAX_ESTUDIANTES];
    float notas[MAX_ESTUDIANTES][NUM_NOTAS];
    int contador;

public:
    Sistema() {
        contador = 0;

        for (int f = 0; f < MAX_ESTUDIANTES; f++) {
            for (int e = 0; e < NUM_NOTAS; e++) {
                notas[f][e] = 0;
            }
        }
    }

    int buscarIndice(int codigo) {
        for (int f = 0; f < contador; f++) {
            if (estudiantes[f].getCodigo() == codigo) {
                return f;
            }
        }
        return -1;
    }

    void registrarEstudiante() {
        if (contador >= MAX_ESTUDIANTES) {
            cout << "No hay espacio.\n";
            return;
        }

        int codigo, edad;
        string nombre;

        cout << "Codigo: ";
        codigo = leerEntero();

        // Validar codigo unico
        if (buscarIndice(codigo) != -1) {
            cout << "El codigo ya existe.\n";
            return;
        }

        cout << "Nombre: ";
        nombre = leerTexto();

        do {
            cout << "Edad: ";
            edad = leerEntero();
            if (edad <= 0) {
                cout << "Edad invalida.\n";
            }
        } while (edad <= 0);

        estudiantes[contador] = Estudiante(codigo, nombre, edad);
        contador++;

        cout << "Estudiante registrado.\n";
    }

    void mostrarEstudiantes() {
        if (contador == 0) {
            cout << "No hay estudiantes.\n";
            return;
        }

        for (int f = 0; f < contador; f++) {
            cout << "\nEstudiante " << f + 1 << ":\n";
            estudiantes[f].mostrar();
        }
    }

    void buscarEstudiante() {
        int codigo;
        cout << "Codigo: ";
        codigo = leerEntero();

        int pos = buscarIndice(codigo);

        if (pos != -1) {
            estudiantes[pos].mostrar();
        } else {
            cout << "No encontrado.\n";
        }
    }

    void asignarNotas() {
        int codigo;
        cout << "Codigo: ";
        codigo = leerEntero();

        int pos = buscarIndice(codigo);

        if (pos != -1) {
            for (int f = 0; f < NUM_NOTAS; f++) {
                float nota;
                do {
                    cout << "Nota " << f + 1 << " (0 - 5): ";
                    nota = leerFloat();

                    if (nota < 0 || nota > 5) {
                        cout << "Nota invalida.\n";
                    }
                } while (nota < 0 || nota > 5);

                notas[pos][f] = nota;
            }
            cout << "Notas guardadas.\n";
        } else {
            cout << "No encontrado.\n";
        }
    }

    void mostrarPromedio() {
        int codigo;
        cout << "Codigo: ";
        codigo = leerEntero();

        int pos = buscarIndice(codigo);

        if (pos != -1) {
            float suma = 0;

            for (int f = 0; f < NUM_NOTAS; f++) {
                suma += notas[pos][f];
            }

            cout << "Promedio: " << suma / NUM_NOTAS << endl;
        } else {
            cout << "No encontrado.\n";
        }
    }

    void mostrarMatriz() {
        if (contador == 0) {
            cout << "No hay datos.\n";
            return;
        }

        for (int f = 0; f < contador; f++) {
            cout << estudiantes[f].getCodigo() << ": ";
            for (int e = 0; e < NUM_NOTAS; e++) {
                cout << notas[f][e] << " ";
            }
            cout << endl;
        }
    }
};


// VISTA

class Vista {
public:
    static void mostrarMenu() {
        cout << "\n=== MENU ===\n";
        cout << "1. Registrar estudiante\n";
        cout << "2. Mostrar estudiantes\n";
        cout << "3. Buscar estudiante\n";
        cout << "4. Asignar notas\n";
        cout << "5. Promedio\n";
        cout << "6. Mostrar notas\n";
        cout << "7. Salir\n";
        cout << "Opcion: ";
    }
};


// MAIN

int main() {
    Sistema sistema;
    int opcion;

    cout << "=== BIENVENIDO AL SISTEMA ===\n";

    do {
        Vista::mostrarMenu();
        opcion = leerEntero();

        switch (opcion) {
            case 1: sistema.registrarEstudiante(); break;
            case 2: sistema.mostrarEstudiantes(); break;
            case 3: sistema.buscarEstudiante(); break;
            case 4: sistema.asignarNotas(); break;
            case 5: sistema.mostrarPromedio(); break;
            case 6: sistema.mostrarMatriz(); break;
            case 7: cout << "Saliendo...\n"; break;
            default: cout << "Opcion invalida\n";
        }

    } while (opcion != 7);

    return 0;
}

File: Corte 2/Python/parcial/parcial.py
MAX_ESTUDIANTES = 30
NUM_NOTAS = 3


# VALIDACIÓN DE DATOS

def leer_entero():
    while True:
        try:
            valor = int(input())
            return valor
        except:
            print("Entrada invalida. Ingrese un numero: ", end="")


def leer_float():
    while True:
        try:
            valor = float(input())
            return valor
        except:
            print("Entrada invalida. Ingrese un numero: ", end="")


def leer_texto():
    while True:
        texto = input()
        if texto.strip() == "":
            print("No puede estar vacio. Intente de nuevo: ", end="")
        else:
            return texto


# ESTUDIANTE

class Estudiante:
    def __init__(self, codigo=0, nombre="", edad=0):
        self.codigo = codigo
        self.nombre = nombre
        self.edad = edad

    def get_codigo(self):
        return self.codigo

    def get_nombre(self):
        return self.nombre

    def get_edad(self):
        return self.edad

    def mostrar(self):
        print(f"Codigo: {self.codigo}")
        print(f"Nombre: {self.nombre}")
        print(f"Edad: {self.edad}")


# SISTEMA

class Sistema:
    def __init__(self):
        self.estudiantes = []
        self.notas = [[0 for _ in range(NUM_NOTAS)] for _ in range(MAX_ESTUDIANTES)]
        self.contador = 0

    def buscar_indice(self, codigo):
        for i in range(self.contador):
            if self.estudiantes[i].get_codigo() == codigo:
                return i
        return -1

    def registrar_estudiante(self):
        if self.contador >= MAX_ESTUDIANTES:
            print("No hay espacio.")
            return

        print("Codigo: ", end="")
        codigo = leer_entero()

        if self.buscar_indice(codigo) != -1:
            print("El codigo ya existe.")
            return

        print("Nombre: ", end="")
        nombre = leer_texto()

        while True:
            print("Edad: ", end="")
            edad = leer_entero()
            if edad > 0:
                break
            print("Edad invalida.")

        estudiante = Estudiante(codigo, nombre, edad)
        self.estudiantes.append(estudiante)
        self.contador += 1

        print("Estudiante registrado.")

    def mostrar_estudiantes(self):
        if self.contador == 0:
            print("No hay estudiantes.")
            return

        for i in range(self.contador):
            print(f"\nEstudiante {i + 1}:")
            self.estudiantes[i].mostrar()

    def buscar_estudiante(self):
        print("Codigo: ", end="")
        codigo = leer_entero()

        pos = self.buscar_indice(codigo)

        if pos != -1:
            self.estudiantes[pos].mostrar()
        else:
            print("No encontrado.")

    def asignar_notas(self):
        print("Codigo: ", end="")
        codigo = leer_entero()

        pos = self.buscar_indice(codigo)

        if pos != -1:
            for i in range(NUM_NOTAS):
                while True:
                    print(f"Nota {i + 1} (0 - 5): ", end="")
                    nota = leer_float()
                    if 0 <= nota <= 5:
                        break
                    print("Nota invalida.")
                self.notas[pos][i] = nota
            print("Notas guardadas.")
        else:
            print("No encontrado.")

    def mostrar_promedio(self):
        print("Codigo: ", end="")
        codigo = leer_entero()

        pos = self.buscar_indice(codigo)

        if pos != -1:
            suma = sum(self.notas[pos])
            print("Promedio:", suma / NUM_NOTAS)
        else:
            print("No encontrado.")

    def mostrar_matriz(self):
        if self.contador == 0:
            print("No hay datos.")
            return

        for i in range(self.contador):
            print(self.estudiantes[i].get_codigo(), end=": ")
            for j in range(NUM_NOTAS):
                print(self.notas[i][j], end=" ")
            print()


# VISTA

class Vista:
    @staticmethod
    def mostrar_menu():
        print("\n=== MENU ===")
        print("1. Registrar estudiante")
        print("2. Mostrar estudiantes")
        print("3. Buscar estudiante")
        print("4. Asignar notas")
        print("5. Promedio")
        print("6. Mostrar notas")
        print("7. Salir")
        print("Opcion: ", end="")


# MAIN

def main():
    sistema = Sistema()

    print("=== BIENVENIDO AL SISTEMA ===")

    while True:
        Vista.mostrar_menu()
        opcion = leer_entero()

        if opcion == 1:
            sistema.registrar_estudiante()
        elif opcion == 2:
            sistema.mostrar_estudiantes()
        elif opcion == 3:
            sistema.buscar_estudiante()
        elif opcion == 4:
            sistema.asignar_notas()
        elif opcion == 5:
            sistema.mostrar_promedio()
        elif opcion == 6:
            sistema.mostrar_matriz()
        elif opcion == 7:
            print("Saliendo...")
            break
        else:
            print("Opcion invalida")


if __name__ == "__main__":
    main()
